data_model/src/error.rs

```diff
@@ -0,0 +1,243 @@
+use parity_scale_codec::{Decode, Encode};
+use thiserror::Error;
+
+use crate::{block::VersionedCommittedBlock, metadata, permission, prelude::*, trigger};
+
+/// Query errors.
+#[derive(Error, Debug, Decode, Encode, IntoSchema)]
+pub enum QueryError {
+    /// Query cannot be decoded.
+    #[error("Query cannot be decoded")]
+    Decode(#[from] Box<iroha_version::error::Error>),
+    /// Query has wrong signature.
+    #[error("Query has the wrong signature: {0}")]
+    Signature(String),
+    /// Query is not allowed.
+    #[error("Query is not allowed: {0}")]
+    Permission(permission::validator::DenialReason),
+    /// Query has wrong expression.
+    #[error("Query has a malformed expression: {0}")]
+    Evaluate(String),
+    /// Query found nothing.
+    #[error("Query found nothing: {0}")]
+    Find(#[from] Box<FindError>),
+    /// Query found wrong type of asset.
+    #[error("Query found wrong type of asset: {0}")]
+    Conversion(String),
+    /// Query without account.
+    #[error("Unauthorized query: account not provided")]
+    Unauthorized,
+}
+
+/// Instruction execution error type
+#[derive(Debug, Error)]
+pub enum IsiError {
+    /// Failed to find some entity
+    #[error("Failed to find. {0}")]
+    Find(#[from] Box<FindError>),
+    /// Failed to assert type
+    #[error("Type assertion failed. {0}")]
+    Type(#[from] TypeError),
+    /// Failed to assert mintability
+    #[error("Mintability violation. {0}")]
+    Mintability(#[from] MintabilityError),
+    /// Failed due to math exception
+    #[error("Math error. {0}")]
+    Math(#[from] MathError),
+    /// Query Error
+    #[error("Query failed. {0}")]
+    Query(#[from] QueryError),
+    /// Metadata Error.
+    #[error("Metadata error: {0}")]
+    Metadata(#[from] metadata::Error),
+    /// Unsupported instruction.
+    #[error("Unsupported {0} instruction")]
+    Unsupported(InstructionType),
+    /// [`FailBox`] error
+    #[error("Execution failed {0}")]
+    FailBox(String),
+    /// Conversion Error
+    #[error("Conversion Error: {0}")]
+    Conversion(String),
+    /// Repeated instruction
+    #[error("Repetition")]
+    Repetition(InstructionType, IdBox),
+    /// Failed to validate.
+    #[error("Failed to validate: {0}")]
+    Validate(#[from] ValidationError),
+}
+
+/// Enumeration of instructions which can have unsupported variants.
+#[derive(Debug, Display, Clone, Copy)]
+pub enum InstructionType {
+    /// Mint
+    Mint,
+    /// Register
+    Register,
+    /// Set key-value pair
+    SetKeyValue,
+    /// Remove key-value pair
+    RemoveKeyValue,
+    /// Grant
+    Grant,
+    /// Transfer
+    Transfer,
+    /// Burn
+    Burn,
+    /// Un-register
+    Unregister,
+    /// Revoke
+    Revoke,
+    /// Set a chain-wide config parameter
+    SetParameter,
+    /// New chain-wide config parameter
+    NewParameter,
+}
+
+/// Type assertion error
+#[derive(Debug, Error, Decode, Encode, IntoSchema)]
+pub enum FindError {
```

Comment:
Given that this is no longer separated from the data by a crate boundary, I fully expect this struct is better generated by a macro.

Reply:
Created #3266

Reply:
Not sure how to implement this as macro because `FindError` contain data from different sources.
Seems like more effort than benefits.